tartiflette_starlette/subscriptions/protocol.py

```diff
@@ -0,0 +1,128 @@
+        await self._send_message(opid, error_type, {"message": str(exception)})
+
+    async def _subscribe(self, opid: str, payload: dict) -> None:
+        stream = self.get_stream(opid, payload)
+        self._operations[opid] = stream
+
+        try:
+            async for item in stream:
+                if opid not in self._operations:
+                    break
+                await self._send_message(opid, optype="data", payload=item)
+        except Exception as exc:  # pylint: disable=broad-except
+            await self._send_error(Exception("Internal error"), opid=opid)
+            raise exc
+
+        await self._send_message(opid, "complete")
+
+    async def _unsubscribe(self, opid: str):
+        operation: typing.AsyncGenerator = self._operations.pop(opid, None)
+        if operation is None:
+            return
+        await operation.aclose()
+
+    # Client message handlers.
+
+    async def _on_connection_init(self, opid: str, **_) -> None:
+        try:
+            await self._send_message(optype=GQL.CONNECTION_ACK)
+        except Exception as exc:  # pylint: disable=broad-except
+            await self._send_error(
+                exc, opid=opid, error_type=GQL.CONNECTION_ERROR
+            )
+            await self.close(1011)
+
+    async def _on_start(self, opid: str, payload: dict) -> None:
+        if opid in self._operations:
+            await self._unsubscribe(opid)
+        await self._subscribe(opid, payload)
+
+    async def _on_stop(self, opid: str, **_) -> None:
+        await self._unsubscribe(opid)
+
+    async def _on_connection_terminate(self, **_) -> None:
+        await self.close(1011)
+
+    # Main task.
+
+    async def _main(self, message: typing.Any):
+        if not isinstance(message, dict):
+            try:
+                message = json.loads(message)
+                if not isinstance(message, dict):
+                    raise TypeError("payload must be a JSON object")
+            except TypeError as exc:
+                await self._send_error(exc)
+
+        optype: str = message.get("type")
+        opid: str = message.get("id")
+        payload: dict = message.get("payload")
+
+        try:
+            handler = getattr(self, f"_on_{optype}", None)
+        except AttributeError:
+            await self._send_error(
+                Exception(f"Unhandled message type: {optype}"), opid=opid
+            )
+        else:
+            await handler(opid=opid, payload=payload)
+
+    # Public API.
+
+    async def on_receive(self, message: dict) -> None:
+        # Subscription execution is a long-lived `async for` operation,
+        # so we must schedule it in a separate task on the event loop.
+        self.schedule(self._main(message))
+
+    async def on_disconnect(self, close_code: int):
+        # NOTE: load keys in list to prevent "size changed during iteration".
+        for opid in list(self._operations):
+            await self._unsubscribe(opid)
```

Comment:
@abusi This is the most transport- and event loop-agnostic version of the protocol I could write.

Do you think it could be worth extracting as a separate lib under the Tartiflette umbrella organization (probably along with the mixins defined in impl.py, except the one about Starlette websockets)? This would allow to reuse the exact same protocol-level logic in the Starlette and aiohttp wrappers.

Maybe a good plan could be releasing this in this package first, then investigate using the protocol and non-Starlette mixins in tartiflette-aiohttp, and from there seeing whether it could work at all.

Reply:
Yep, exactly what I was thinking, let's release this, and then it we'll be easier to make the protocol an external lib that ttftt-aiohttp, ttftt-starlette and others could rely on.
This is great work.